File: app/src/main/java/com/asfoundation/wallet/interact/FindDefaultWalletInteract.java
package com.asfoundation.wallet.interact;

import com.asfoundation.wallet.entity.Wallet;
import com.asfoundation.wallet.repository.WalletRepositoryType;
import io.reactivex.Single;

public class FindDefaultWalletInteract {

  private final WalletRepositoryType walletRepository;

  public FindDefaultWalletInteract(WalletRepositoryType walletRepository) {
    this.walletRepository = walletRepository;
  }

  public Single<Wallet> find() {
    return walletRepository.getDefaultWallet()
        .onErrorResumeNext(throwable -> walletRepository.fetchWallets()
            .filter(wallets -> wallets.length > 0)
            .map(wallets -> wallets[0])
            .flatMapCompletable(walletRepository::setDefaultWallet)
            .andThen(walletRepository.getDefaultWallet()));
  }
}

File: app/src/main/java/com/asfoundation/wallet/interact/FindDefaultWalletInteract.kt
package com.asfoundation.wallet.interact

import com.asfoundation.wallet.entity.Wallet
import com.asfoundation.wallet.repository.WalletRepositoryType
import io.reactivex.Observable
import io.reactivex.Scheduler
import io.reactivex.Single

class FindDefaultWalletInteract(private val walletRepository: WalletRepositoryType,
                                private val scheduler: Scheduler) {
  fun find(): Single<Wallet> {
    return walletRepository.getDefaultWallet()
        .subscribeOn(scheduler)
        .onErrorResumeNext {
          walletRepository.fetchWallets()
              .filter { wallets -> wallets.isNotEmpty() }
              .map { wallets: Array<Wallet> ->
                wallets[0]
              }
              .flatMapCompletable { wallet: Wallet ->
                walletRepository.setDefaultWallet(wallet.address)
              }
              .andThen(walletRepository.getDefaultWallet())
        }
  }

  fun observe(): Observable<Wallet> {
    return walletRepository.observeDefaultWallet()
        .subscribeOn(scheduler)
        .onErrorResumeNext { _: Throwable ->
          return@onErrorResumeNext walletRepository.fetchWallets()
              .filter { wallets -> wallets.isNotEmpty() }
              .map { wallets: Array<Wallet> ->
                wallets[0]
              }
              .flatMapCompletable { wallet: Wallet ->
                walletRepository.setDefaultWallet(wallet.address)
              }
              .andThen(walletRepository.observeDefaultWallet())
        }
  }
}